packages/react/src/hooks/contracts/usePrepareContractWrite.ts

```diff
@@ -0,0 +1,136 @@
+  PrepareWriteContractConfig,
+  'signerOrProvider'
+>
+export type UsePrepareContractWriteConfig = QueryConfig<
+  PrepareWriteContractResult,
+  Error
+>
+
+export const queryKey = (
+  {
+    args,
+    addressOrName,
+    contractInterface,
+    functionName,
+    overrides,
+  }: UsePrepareContractWriteArgs,
+  { chainId, signer }: { chainId?: number; signer?: FetchSignerResult },
+) =>
+  [
+    {
+      entity: 'prepareContractTransaction',
+      addressOrName,
+      args,
+      chainId,
+      contractInterface,
+      functionName,
+      overrides,
+      signer,
+    },
+  ] as const
+
+const queryKeyHashFn = ([queryKey_]: ReturnType<typeof queryKey>) => {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { contractInterface, signer, ...rest } = queryKey_
+  return hashQueryKey([rest, signer?._address])
+}
+
+const queryFn =
+  ({ signer }: { signer?: FetchSignerResult }) =>
+  ({
+    queryKey: [
+      { args, addressOrName, contractInterface, functionName, overrides },
+    ],
+  }: QueryFunctionArgs<typeof queryKey>) => {
+    return prepareWriteContract({
+      args,
+      addressOrName,
+      contractInterface,
+      functionName,
+      overrides,
+      signer,
+    })
+  }
+
+/**
+ * @description Hook for preparing a contract write to be sent via [`useContractWrite`](/docs/hooks/useContractWrite).
+ *
+ * Eagerly fetches the parameters required for sending a contract write transaction such as the gas estimate.
+ *
+ * @example
+ * import { useContractWrite, usePrepareContractWrite } from 'wagmi'
+ *
+ * const { config } = usePrepareContractWrite({
+ *  addressOrName: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
+ *  contractInterface: wagmigotchiABI,
+ *  functionName: 'feed',
+ * })
+ * const { data, isLoading, isSuccess, write } = useContractWrite(config)
+ *
+ */
+export function usePrepareContractWrite({
+  addressOrName,
+  contractInterface,
+  functionName,
+  args,
+  overrides,
+  cacheTime,
+  enabled = true,
+  staleTime,
+  suspense,
+  onError,
+  onSettled,
+  onSuccess,
+}: UsePrepareContractWriteArgs & UsePrepareContractWriteConfig) {
+  const chainId = useChainId()
+  const { data: signer } = useSigner()
+
+  const prepareContractWriteQuery = useQuery(
+    queryKey(
+      {
+        addressOrName,
+        contractInterface,
+        functionName,
+        args,
+        overrides,
+      },
+      { chainId, signer },
+    ),
+    queryFn({ signer }),
+    {
+      cacheTime,
+      enabled: Boolean(enabled && signer),
+      queryKeyHashFn,
+      staleTime,
+      suspense,
+      onError,
+      onSettled,
+      onSuccess,
+    },
+  )
+  return Object.assign(prepareContractWriteQuery, {
+    config: {
+      addressOrName,
+      args,
+      contractInterface,
+      overrides,
+      functionName,
+      request: undefined,
+      mode: 'prepared',
+      ...prepareContractWriteQuery.data,
+    } as PrepareWriteContractResult,
+  })
```

Comment:
How does this affect the tracked behavior of `useQuery`?

Reply:
Still works for the stuff in `prepareContractWriteQuery`, however, we lose it for `config`. I figured most usages of `usePrepareContractWrite` will have `config` in the return value, so not sure if it was worth the effort to tweak this into our custom `useQuery`.